### PR TITLE
[FX][Dynamo] Add tests for nn_module_stack and source_fn_stack metadata contracts

### PR DESCRIPTION
**Pull request by lahead**
## Summary

Adds targeted tests for the `nn_module_stack` and `source_fn_stack` node
metadata that FX tracing and Dynamo attach to graph nodes.  These metadata
are consumed by multiple downstream systems (ONNX export, quantisation,
torch.export serialisation, custom compiler backends), but the existing
test coverage only validates basic 2-level nesting with symbolic_trace
and checks for metadata *presence* rather than *structure*.

This PR adds 8 tests that lock down the metadata contracts:

**test/test_fx.py** (symbolic_trace path):
- `test_nn_module_stack_deep_nesting` — dynamically tests depths 3–9 via `subTest`, verifying all intermediate levels are fully recorded and ordered outermost→innermost.
- `test_nn_module_stack_module_list` — `nn.ModuleList` children appear with numeric path components (`blocks.0`, `blocks.1`, …).
- `test_nn_module_stack_class_entries_are_types` — class entries in `nn_module_stack` are real types (not strings) for both user-defined and built-in modules.

**test/dynamo/test_export.py** (Dynamo export path):
- `test_export_nn_module_stack_contains_module_path_prefix` — Dynamo paths use the `L['self'].` prefix format.
- `test_export_nn_module_stack_class_is_type` — class entries remain real `type` objects (not serialised strings) with `__name__`.
- `test_export_nn_module_stack_nested_module_list` — `ModuleList` numeric indices appear in Dynamo-prefixed paths.
- `test_export_nn_module_stack_ordering_outermost_to_innermost` — dynamically tests depths 3–9 via `subTest`, verifying dict iteration order is outermost module first and the deepest stack reflects the full nesting chain.
- `test_export_source_fn_stack_entries_are_named_callable_pairs` — validates the `list[tuple[str, callable]]` schema where `entry[1]` carries `__name__`.

## Motivation

Several components rely on undocumented structural invariants of these
metadata fields:
- **ONNX exporter** reads `nn_module_stack` to map FX nodes to ONNX node
  metadata.
- **Quantisation** (`torch.ao.quantization.pt2e.qat_utils`) inspects
  `source_fn_stack` class entries to identify BatchNorm operations.
- **torch.export** normalises `L['self'].` prefixes during serialisation
  (`_normalize_nn_module_stack` in `torch/export/_trace.py`).
- **Custom compiler backends** reconstruct module hierarchies from
  `nn_module_stack` to map operations back to user-written module
  structure.

Without explicit tests, changes to the tracing infrastructure can silently
break these consumers.  These tests make the implicit contracts explicit.

## Test Plan

All 8 new tests pass locally:

```
python test/test_fx.py 
TestFX.test_nn_module_stack_deep_nesting
TestFX.test_nn_module_stack_module_list
TestFX.test_nn_module_stack_class_entries_are_types

python test/dynamo/test_export.py 
ExportTests.test_export_nn_module_stack_contains_module_path_prefix
ExportTests.test_export_nn_module_stack_class_is_type
ExportTests.test_export_nn_module_stack_nested_module_list
ExportTests.test_export_nn_module_stack_ordering_outermost_to_innermost
ExportTests.test_export_source_fn_stack_entries_are_named_callable_pairs
```

Existing tests unaffected (no production code changes).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude explain those CI failures, check if they are related to this PR, and review the changes

**Comment by ezyang** [comment]
@claude do you think potentially these tests are brittle?

**Comment by lahead** [comment]
Applied the review suggestions partially - reduced depth iterations to (3, 6, 9) in both files, and replaced the dot-count ordering with a prefix-chain check in test_export.py, kept the dot-count monotonicity check in test_fx.py though as even if module name contains a dot the check is still valid because it's not expecting a specific number of dots.

On the L['self']. prefix: kept it because it's the stable format that normalize_nn_module_stack, inductor's clean_stack_name, and the ONNX exporter all already hardcode against it explicitly.

**Comment by lahead** [comment]
Hi @jansel, please have a look if this change can be merged

**Comment by jansel** [comment]
@claude look at this again and update your opinion on brittleness

**Comment by lahead** [comment]
Thanks @jansel for additional feedback. I pushed fixes for both test_export_source_fn_stack_entries_are_named_callable_pairs and test_export_nn_module_stack_contains_module_path_prefix as suggested. 

Regarding the test_nn_module_stack_deep_nesting I would suggest to stay with with exact equality check. The deepest node’s nn_module_stack is expected to contain exactly the module chain traversed by that node, in outermost-to-innermost order. Exact equality verifies completeness, ordering, and absence of unexpected or duplicate entries in one assertion. The suggested change to a subset check would make it allow extra metadata entries and would not detect an incorrectly attributed module.
